File: sdk/src/grants.ts
/**
 * The grant calls after connect: refresh, revoke, status (SPEC §7). Kept
 * apart from index.ts so the fetch wrapper can ask status() without a
 * circular import. A connection that has ended answers as a
 * ConnectionEndedError from status() and refresh() (ended.ts).
 */
import { assertVaultGrant } from "./direct.js";
import { endedFromInfo, endedFromVault } from "./ended.js";
import { DEFAULT_BASE_URL, api, type Auth } from "./http.js";
import { GrantInfo, OutletError, OutletSession, RequestOptions } from "./types.js";

/** Confidential clients authenticate with the app secret; public clients
 *  (PKCE, SPEC §7.1) carry a grant-scoped refresh token instead. */
export function authOf(opts: RequestOptions): Auth {
  return opts.refreshToken
    ? { bearer: opts.refreshToken }
    : { appSecret: opts.appSecret };
}

/** A vault refusal that means the connection ended becomes the typed
 *  error; every other refusal passes through as it was. */
function surfaced(e: unknown, grantId: string): unknown {
  return (e instanceof OutletError && endedFromVault(e, grantId)) || e;
}

/**
 * Re-fetch (and possibly rotate) keys for an existing grant. Server-side
 * for confidential clients. Public clients (RequestOptions.refreshToken)
 * get a ROTATED token back (OAuth 2.1 §6.1): the token they presented is
 * void the moment the vault answers, so callers MUST persist the returned
 * refreshToken before making another call — the old one now 401s.
 * Throws ConnectionEndedError when the connection is capped, revoked, or
 * the token no longer refreshes it.
 */
export async function refresh(
  grantId: string,
  opts: RequestOptions = {},
): Promise<OutletSession & { refreshToken?: string }> {
  assertVaultGrant(grantId);
  let r: OutletSession & { refresh_token?: string };
  try {
    r = await api<OutletSession & { refresh_token?: string }>(
      opts.baseUrl ?? DEFAULT_BASE_URL,
      `/grants/${grantId}/refresh`, { method: "POST" }, authOf(opts));
  } catch (e) {
    throw surfaced(e, grantId);
  }
  // why the rename: the wire is snake_case (SPEC §7.1); the SDK surface is
  // camelCase. Dropping the rotated token here stranded public clients —
  // their stored token was already voided server-side (security review).
  const { refresh_token, ...session } = r;
  return refresh_token ? { ...session, refreshToken: refresh_token } : session;
}

/** App-initiated revocation (users can always revoke from their dashboard). */
export async function revoke(
  grantId: string,
  opts: RequestOptions = {},
): Promise<void> {
  assertVaultGrant(grantId);
  await api<unknown>(opts.baseUrl ?? DEFAULT_BASE_URL,
    `/grants/${grantId}`, { method: "DELETE" }, authOf(opts));
}

/**
 * Current status + spend for a grant. A pure read: the vault's /status
 * endpoint never returns key material (the old GET /grants/:id habit
 * re-delivered the full key on every status poll — security review).
 * spendUsd is the vault's last meter reading: month-to-date, up to ~5
 * minutes stale, 0 for a grant not yet metered. A connection that has
 * ended throws ConnectionEndedError instead (its `info` holds the answer),
 * so an app never reads a capped or revoked connection as one to keep
 * calling with.
 */
export async function status(
  grantId: string,
  opts: RequestOptions = {},
): Promise<GrantInfo> {
  assertVaultGrant(grantId);
  let info: GrantInfo;
  try {
    info = await api<GrantInfo>(opts.baseUrl ?? DEFAULT_BASE_URL,
      `/grants/${grantId}/status`, undefined, authOf(opts));
  } catch (e) {
    throw surfaced(e, grantId);
  }
  const end = endedFromInfo(info);
  if (end) throw end;
  return info;
}

File: sdk-python/src/useoutlet/grants.py
"""The grant calls after connect: refresh, revoke, status (SPEC section 7).
Module-level functions that take the credential as arguments; the Outlet
client wraps them and holds the refresh token for you. A connection that has
ended answers as a ConnectionEndedError from status() and refresh() (ended.py)."""

from __future__ import annotations

from typing import NoReturn

from . import http as _http
from .direct import assert_vault_grant
from .ended import ended_from_info, ended_from_vault
from .http import DEFAULT_TIMEOUT, Auth, api
from .types import GrantInfo, OutletError, OutletSession


def auth_of(app_secret: str | None, refresh_token: str | None) -> Auth:
    """Confidential clients authenticate with the app secret; public clients
    (PKCE, SPEC section 7.1) carry a grant-scoped refresh token instead."""
    return Auth(bearer=refresh_token) if refresh_token else Auth(app_secret=app_secret)


def _raise_surfaced(e: OutletError, grant_id: str) -> NoReturn:
    """A vault refusal that means the connection ended becomes the typed error;
    every other refusal passes through as it was."""
    end = ended_from_vault(e, grant_id)
    if end is not None:
        raise end from e
    raise e


def refresh(
    grant_id: str,
    *,
    app_secret: str | None = None,
    refresh_token: str | None = None,
    base_url: str | None = None,
    timeout: float = DEFAULT_TIMEOUT,
) -> OutletSession:
    """Re-fetch (and possibly rotate) keys for an existing grant. Public clients
    (refresh_token) get a ROTATED token back (OAuth 2.1 section 6.1): the token
    they presented is void the moment the vault answers, so persist the
    returned refresh_token before making another call. The old one now 401s.
    Raises ConnectionEndedError when the connection is capped, revoked, or the
    token no longer refreshes it."""
    assert_vault_grant(grant_id)
    try:
        r = api(
            base_url or _http.DEFAULT_BASE_URL,
            f"/grants/{grant_id}/refresh",
            method="POST",
            auth=auth_of(app_secret, refresh_token),
            timeout=timeout,
        )
    except OutletError as e:
        _raise_surfaced(e, grant_id)
    return OutletSession.from_wire(r)


def revoke(
    grant_id: str,
    *,
    app_secret: str | None = None,
    refresh_token: str | None = None,
    base_url: str | None = None,
    timeout: float = DEFAULT_TIMEOUT,
) -> None:
    """App-initiated revocation (users can always revoke from their account page)."""
    assert_vault_grant(grant_id)
    api(
        base_url or _http.DEFAULT_BASE_URL,
        f"/grants/{grant_id}",
        method="DELETE",
        auth=auth_of(app_secret, refresh_token),
        timeout=timeout,
    )


def status(
    grant_id: str,
    *,
    app_secret: str | None = None,
    refresh_token: str | None = None,
    base_url: str | None = None,
    timeout: float = DEFAULT_TIMEOUT,
) -> GrantInfo:
    """Current status and spend for a grant. A pure read: the vault's /status
    endpoint never returns key material. spend_usd is the vault's last meter
    reading: month-to-date, up to about five minutes stale, 0 for a grant not
    yet metered. A connection that has ended raises ConnectionEndedError
    instead (its `info` holds the answer), so an app never reads a capped or
    revoked connection as one to keep calling with."""
    assert_vault_grant(grant_id)
    try:
        r = api(
            base_url or _http.DEFAULT_BASE_URL,
            f"/grants/{grant_id}/status",
            auth=auth_of(app_secret, refresh_token),
            timeout=timeout,
        )
    except OutletError as e:
        _raise_surfaced(e, grant_id)
    info = GrantInfo.from_wire(r)
    end = ended_from_info(info)
    if end is not None:
        raise end
    return info
